clients/drcachesim/tools/invariant_checker.cpp

```diff
@@ -837,13 +850,24 @@ invariant_checker_t::check_for_pc_discontinuity(
     std::string error_msg = "";
     bool have_cond_branch_target = false;
     addr_t cond_branch_target = 0;
-    const addr_t prev_instr_trace_pc = shard->prev_instr_.instr.addr;
-    if (prev_instr_trace_pc != 0 /*first*/ &&
+    memref_t prev_instr = shard->prev_instr_;
+    bool memref_is_kernel_event_marker = false;
+#ifdef UNIX
+    // Identify whether the current memref is a marker instead of an instruction. This
+    // will be handled differently.
+    if (memref.marker.type == TRACE_TYPE_MARKER &&
+        memref.marker.marker_type == TRACE_MARKER_TYPE_KERNEL_EVENT) {
+        memref_is_kernel_event_marker = true;
+        prev_instr = shard->last_instr_in_cur_context_;
+    }
+#endif
+
+    if (prev_instr.instr.addr != 0 /*first*/ &&
         // We do not bother to support legacy traces without encodings.
         expect_encoding && type_is_instr_direct_branch(shard->prev_instr_.instr.type)) {
         if (shard->prev_instr_.instr.encoding_is_new)
-            shard->branch_target_cache.erase(prev_instr_trace_pc);
-        auto cached = shard->branch_target_cache.find(prev_instr_trace_pc);
+            shard->branch_target_cache.erase(prev_instr.instr.addr);
+        auto cached = shard->branch_target_cache.find(prev_instr.instr.addr);
```

Comment:
I had to go add a call to check_for_pc_discontinuity at a kernel event to my code for #6213 and #5490 which overlaps with this PR.  One issue I was about to fix was shard->prev_instr_decoded_ but I figured I should avoid too much overlap -- but I don't see it being fixed here.  Here is the TODO I have in my code:

            // TODO i#5912: This prev_instr_decoded_ isn't always the same as the
            // decoding of the passed-in prev_instr, for signals.  What we should do is
            // create a new instr_info_t struct with fields for the memref_t plus the
            // from-decoding attributes is_syscall, writes_memory, and branch_target, and
            // fill it in at decode time.  We can then store it as prev_instr_ and in
            // pre_signal_instr and last_instr_in_cur_context_ and pass it here as
            // prev_instr.  We can then delete this branch_target_cache and
            // prev_instr_decoded_.  For now we live with this inaccuracy as it is
            // rare to not match and we leave the cleanup for i#5912.

Reply:
I'm not sure I understand why this branch_target_cache was added in the first place: we always had prev_instr_decoded so we always had the direct target right there.  What we want is a cache of the decoding to avoid calling decode_from_copy() on every single instruction: that's the expensive step.

Reply:
> What we want is a cache of the decoding to avoid calling decode_from_copy() on every single instruction:

Yes we should add that now; it'd require updating our current logic of free-ing the `instr_t`s too. I'm a little concerned about memory usage going up due to tracking many `instr_t`, but we can test that experimentally on a large trace.